docs: add TIP-1042 FeeAMM TIP-403 policy exemptions (#3425)
tempoxyz/tempo: tips/tip-1042.md

## Assumptions

- The FeeManager address (`0xfeec000000000000000000000000000000000000`) is a protocol-level singleton that cannot be controlled by any external account.
- Token issuers understand that whitelisting the FeeManager address opts their token into AMM participation (pool creation, rebalancing, LP operations).

---

# Specification

## Overview

The changes are divided into two categories:

1. **Fee collection path**: Remove TIP-403 checks on the FeeManager address during protocol-initiated fee operations.
2. **Public AMM operations**: Preserve existing TIP-403 enforcement, plus add a new authorization gate on `mint`.

## Fee Collection Path — Exempt FeeManager

The following operations skip TIP-403 authorization checks for the FeeManager address. The counterparty (user, validator) is still checked.

> ### `collectFeePreTx`

Reviewer: Can we add a similar section explicitly specifying changes for `collectFeePostTx`
i am assuming there will be no change here, because we don't do any checks atm.

Currently checks both the user (as sender) and the FeeManager (as recipient):

```solidity
uint64 policyId = ITIP20(userToken).transferPolicyId();
if (
    !TIP403_REGISTRY.isAuthorizedSender(policyId, user)
        || !TIP403_REGISTRY.isAuthorizedRecipient(policyId, address(this))
) {
    revert ITIP20.PolicyForbids();
}
```

**Change**: Remove the `isAuthorizedRecipient(policyId, address(this))` check. Only check the user:

```solidity
uint64 policyId = ITIP20(userToken).transferPolicyId();
if (!TIP403_REGISTRY.isAuthorizedSender(policyId, user)) {
    revert ITIP20.PolicyForbids();
}
```

### `transferFeePreTx`

No change. This function already does not perform TIP-403 checks (it relies on the caller having checked).

### `executeFeeSwap`

No change. This is an internal function that only updates pool reserves.

### `transferFeePostTx`

No change. This function already does not perform TIP-403 checks.

## Public AMM Operations — Full TIP-403 Enforcement

The following operations continue to enforce TIP-403 on all participants, including the FeeManager address, via the underlying `transfer()` and `systemTransferFrom()` calls on TIP-20 tokens.

### `distributeFees`

> No change. Calls `IERC20(token).transfer(validator, amount)`, which enforces TIP-403 on both sender (FeeManager) and recipient (validator).

Reviewer 2: nit
```suggestion
No change. Calls `ITIP20(token).transfer(validator, amount)`, which enforces TIP-403 on both sender (FeeManager) and recipient (validator).
```

### `burn`

> No change. Calls `IERC20(userToken).transfer(to, amountUserToken)` and `IERC20(validatorToken).transfer(to, amountValidatorToken)`, which enforce TIP-403 on both sender (FeeManager) and recipient (`to`).

Reviewer 2: ```suggestion
No change. Calls `ITIP20(userToken).transfer(to, amountUserToken)` and `ITIP20(validatorToken).transfer(to, amountValidatorToken)`, which enforce TIP-403 on both sender (FeeManager) and recipient (`to`).
```

### `rebalanceSwap`

No change. Calls `systemTransferFrom(msg.sender, address(this), amountIn)` and `transfer(to, amountOut)`, which enforce TIP-403 on all participants.

### `mint` — Enhanced Authorization

In addition to the existing TIP-403 enforcement on the `validatorToken` transfer, `mint` adds an authorization check on the `userToken` for the caller.

> **Rationale**: The caller is providing liquidity to a pool that facilitates swaps for `userToken`. By minting, they become an economic participant in that token's ecosystem — pool reserves shift during fee swaps and rebalancing, and on `burn` the LP receives pro-rata shares of both tokens. The caller should be authorized to interact with the `userToken` even though no `userToken` moves during `mint`.

Reviewer 3: The new sender check on `msg.sender` is bypassable.

Mint only checks the `msg.sender`, rather than the `to` address. Burn only checks `to`, and not the LP. Either gate can be passed by routing through a whitelisted helper, neither side of the lifecycle actually pins authorization to the LP itself.

Reviewer 4: I think makes sense to check `to` as a recipient on `mint` and as a sender on `burn` (on the user token and the validator token), roughly consistent with what we do on the DEX.

> **Change**: Before processing the mint, check that the caller is authorized as a sender on the `userToken` and that the FeeManager is an authorized recipient — as if the caller were transferring `userToken` into the pool:

Reviewer: Nice, we can use `ensure_transfer_authorized` here directly

```solidity
function mint(
    address userToken,
    address validatorToken,
    uint256 amountValidatorToken,
    address to
) external returns (uint256 liquidity) {
    // Existing validations
    if (userToken == validatorToken) revert IdenticalAddresses();
    if (amountValidatorToken == 0) revert InvalidAmount();
    _requireUSDTIP20(userToken);
    _requireUSDTIP20(validatorToken);

    // New: treat mint as if caller were transferring userToken to the pool.
    // Caller must be authorized as sender, FeeManager as recipient.
    uint64 userTokenPolicyId = ITIP20(userToken).transferPolicyId();
    if (
        !TIP403_REGISTRY.isAuthorizedSender(userTokenPolicyId, msg.sender)

>             || !TIP403_REGISTRY.isAuthorizedRecipient(userTokenPolicyId, address(this))

Reviewer 2: `to` is not checked as an authorized LP-share recipient, should we add here `|| !TIP403_REGISTRY.isAuthorizedRecipient(userTokenPolicyId, to) || !TIP403_REGISTRY.isAuthorizedRecipient(ITIP20(validatorToken).transferPolicyId, to)`

Reviewer 2: following up from https://github.com/tempoxyz/tempo/pull/3425/changes#r3150523058

    ) {
        revert ITIP20.PolicyForbids();
    }

>     // ... rest of mint logic (unchanged)

Reviewer 2: should we add here an explicit `systemTransferFrom(msg.sender, address(this), amountValidatorToken)` line? this is the call that enforces the validatorToken policy as per the Behavioral Summary "FeeManager checked as recipient (both tokens) / Caller checked as sender (both tokens)"

Reviewer 2: it is technically part of "rest of mint logic (unchanged)" but maybe worth mentioned as we do for `rebalanceSwap` https://github.com/tempoxyz/tempo/pull/3425/changes#diff-96299f1d34b0d26567f7873b97e841bc32f07a9fb1e3761f6c522f932b025803R91

}
```

> Note: There is an inherent TOCTOU (time-of-check-time-of-use) gap — a policy may change between when the LP mints and when the pool is used for fee swaps or rebalancing. This is a best-effort check and is consistent with how TIP-403 operates elsewhere (e.g., a token holder may be blacklisted after receiving tokens).

Reviewer 2: per invariant 1, fee swaps don't check the FeeManager
```suggestion
Note: There is an inherent TOCTOU (time-of-check-time-of-use) gap — a policy may change between when the LP mints and when the pool is used for subsequent `rebalanceSwap`, `burn`, or `distributeFees`. This is a best-effort check and is consistent with how TIP-403 operates elsewhere (e.g., a token holder may be blacklisted after receiving tokens).
```

> Additionally, since `rebalanceSwap` checks the relevant policies, a token admin can prevent the liquidity on the FeeAMM from being refreshed, which means it will eventually be exhausted. This should give token admins an adequate tool for disabling the FeeAMM for their token even if someone manages to create liquidity on the pool before the admin sets that policy.

Reviewer 3: The spec enforces TIP-403 on FeeManager as sender for `burn`, `distributeFees`, and `rebalanceSwap`. If an issuer revokes `FeeManager` from the whitelist after LPs have minted or fees have been collected, those outflows all revert. 

The spec acknowledges this but another option could be exempting the `FeeManager` from TIP-403 on the outflow paths (burn, distributeFees, rebalanceSwap) in addition to fee collection. The mint gate could stay as an issuer's opt-in for AMM participation (or we could consider adding a mapping for fee token enabled/disabled).  

We should consider an approach where once funds are in the pool, they arent able to be frozen due to TIP-403 checks against the `FeeManager`. Existing checks against the user should remain unchanged as the spec suggests.

| `transferFeePostTx` | N/A (no check today) | N/A | — |
| `distributeFees` | ✅ FeeManager checked as sender | ✅ Validator checked as recipient | — |
| `mint` | ✅ FeeManager checked as recipient (both tokens) | ✅ Caller checked as sender (both tokens) | ✅ (see above) |
| `burn` | ✅ FeeManager checked as sender | ✅ `to` checked as recipient | — |
| `rebalanceSwap` | ✅ FeeManager checked as recipient + sender | ✅ Caller + `to` checked | — |

## Issuer Control Surface

The FeeManager address exemption does **not** remove issuer control. Instead, it shifts the control mechanism from "whitelist the FeeManager address" to "whitelist determines AMM participation":

| Issuer action | Fee payment | AMM pools | `distributeFees` |
|---------------|-------------|-----------|-------------------|
| Whitelist FeeManager | ✅ | ✅ Pools can be created, rebalanced | ✅ Fees distributed |
| Don't whitelist FeeManager | ✅ Same-token fees work. Cross-token fees require pool liquidity (see below). | ❌ `mint` blocked (FeeManager not authorized as recipient on either token). No pools → no cross-token fee swaps. | ❌ `transfer` from FeeManager blocked |

### Edge Case: Stranded Fees

> If a token does not whitelist the FeeManager address, same-token fees (where `userToken == validatorToken`) will be collected successfully — the fee collection path is exempt. However, `distributeFees` calls `transfer()` which checks the FeeManager as sender. If the FeeManager is not whitelisted, the validator cannot claim these fees and they remain stranded in the FeeManager.

Reviewer 3: Its also possible to hit the case where LP funds are frozen if the token issuer initially whitelists (or doesnt blacklist) the fee token, users LP into the pool and then the token issuer removes the `FeeManager` from the whitelist. Currently there is no recovery mechanism for this case, once `collectedFees[v][x]` accrues for a non-whitelisted token, the value is stuck.

As [mentioned above](https://github.com/tempoxyz/tempo/pull/3425#discussion_r3118723593), this would be fixed by exempting `burn` from `TIP-403` checks.

Reviewer 4: Yeah, I think could make sense to pave the path where the fee token can turn off the fee manager and put it into "withdraw only" mode.

In that case, we should skip the check on whether FeeManager is an authorized sender on `burn` and `distributeFees` but not the checks on `rebalanceSwap`.

Reviewer 4: This would mean that in the non-pool case, there's no way to prevent fees from being paid in the token to validators who set it as their fee token (other than by blacklisting validators). But I think this is OK? The FeeManager is acting as a passthrough. The token can blacklist the validator

Reviewer 4: Eh, on thinking about it more, I think it's OK if blacklisting it as the sender causes funds to be stuck inside; that's ordinary behavior for a contract. If a token wants to put it in withdraw-only mode, it can set a compound policy
